File: frequencyTopic2020/backTrack/exist.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int h;
    int w;

    bool exist(vector<vector<char>> &board, string word) {
        if (board.empty() || board[0].empty()) {
            return false;
        }
        h = board.size();
        w = board[0].size();
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (backtrack(board, i, j, word, 0)) {
                    return true;
                }
            }
        }
        return false;
    }

    bool backtrack(vector<vector<char>> &board, int i, int j, string word, int idx) {
        if (board[i][j] != word[idx]) {
            return false;
        }
        if (idx == word.size() - 1) {
            return true;
        }
        board[i][j] = '0';
        bool res = fales;
        if (i - 1 >= 0 && board[i - 1][j] != '0') {
            res || = backtrack(board, i - 1, j, word, idx + 1);
        }
        if (i + 1 < h && board[i + 1][j] != '0') {
            res || = backtrack(board, i + 1, j, word, idx + 1);
        }
        if (j - 1 >= 0 && board[i][j - 1] != '0') {
            res || = backtrack(board, i, j - 1, word, idx + 1);
        }
        if (j + 1 < w && board[i][j + 1] != '0') {
            res || = backtrack(board, i, j + 1, word, idx + 1);
        }
        board[i][j] = word[idx];
        return res;
    }
};

File: swordOffer/go/dfs/exist.go
package main

func exist(board [][]byte, word string) bool {

	if len(board) == 0 || len(board[0]) == 0 {
		return false
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[0]); j++ {
			if backtrack(board, i, j, word, 0) {
				return true
			}
		}
	}

	return false
}

func backtrack(board [][]byte, x, y int, word string, idx int) bool {

	if idx >= len(word) || board[x][y] != word[idx] {
		return false
	}
	if idx == len(word)-1 {
		return true
	}
	board[x][y] = '0'

	var ok bool
	if x-1 >= 0 && board[x-1][y] != '0' {
		ok = ok || backtrack(board, x-1, y, word, idx+1)
	}
	if x+1 < len(board) && board[x+1][y] != '0' {
		ok = ok || backtrack(board, x+1, y, word, idx+1)
	}
	if y-1 >= 0 && board[x][y-1] != '0' {
		ok = ok || backtrack(board, x, y-1, word, idx+1)
	}
	if y+1 < len(board[0]) && board[x][y+1] != '0' {
		ok = ok || backtrack(board, x, y+1, word, idx+1)
	}
	board[x][y] = word[idx]

	return ok
}
